File: etc/크루스칼 알고리즘/크루스칼.py
def find_parent(parent, x): #특정 원소가 속한 집합 찾기
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]
    
def union_parent(parent, a, b): #두 원소가 속한 집합 찾기
    a = find_parent(a)
    b = find_parent(b)
    if a< b:
        parent[b] = a
    else:
        parent[a] = b
        
 v, e = map(int, input().split())
 parent = [0] * (v+1)
 
 edges = []
 result  = 0
 
 for i in range ( 1, v+1):
     parent[i] = i
     
 for _ in range(e):
     a, b, cost = map(int, input().split())
     edges.append((cost, a, b)) #비용순 정렬을 위해 첫 번째 원소를 비용으로 설정

edges.sort()

for edge in edges: #간선을 하나씩 확인
    cost , a, b = edge
    
    if find_parent(parent, a) != find_parent(parent, b): #사이클이 발생하지 않는 경우에만 집합 포함
        union_parent(parent, a, b)
        result+=cost
        
print(result) #간선 비용의 합

File: etc/크루스칼 알고리즘/크루스칼.cpp
int v, e;
vector<pair<int, pair<int, int>> edges;
int result = 0;

int findParent(int x){
    if(x == parent[x]) return x;
    return parent[x] = findParent(parent[x]);
}

void unionParent(int a, int b){
    a = findParent(a);
    b = findParent(b);
    if (a<b) parent[b] = a;
    else parent[a] = b;
}
int main(){
    for(int i=1 ;i <=v; i++)
        parent[i] = i;
    for(int i=0; i < e; i++){
        int a, b, cost;
        cin >> a>> b>> cost;
        edges.push_back({cost, {a, b}});
    }
    
    sort(edges.begin(), edges.end());
    
    for(int i=0; i < edges.size(); i++){
        int cost = edges[i].first;
        int a = edges[i].second.first;
        int b = edges[i].second.second;
        
        if(findParent(a) != findParnet(b)){ //사이클 발생 x경우만
            unionParent(a, b);
            result += cost;
        }
    }
    cout << result;
}
        
    
